Update DAIInterestRateModel to v4 (fixes #230)

fmorisan opened the pull request:
This patch changes the DAIInterestRateModel contract to use a `SECONDS_PER_BLOCK` parameter of `12`, reflecting post-merge protocol changes. It also updates the IRM to look at the `ETH-B` stability fee in order to calculate the borrow rate, instead of using the more conservative `ETH-A` as a reference.

This fixes #230 and enables reactivation of the Dai Savings Rate for cDAI holders, once the contract is deployed and the IRM for cDAI is updated to the new version.

Related governance forum thread: https://www.comp.xyz/t/compound-dsr-proposal/3856

fmorisan committed:
Update DAIInterestRateModel to v4 (#230)

MathisGD commented on contracts/DAIInterestRateModelV4.sol:
```diff
@@ -0,0 +1,132 @@
+/**
+  * @title Compound's DAIInterestRateModel Contract (version 4)
+  * @author Compound, Dharma (modified by Maker Growth)
+  * @notice The change from v3 to v4 of this contract was just the `SECONDS_PER_BLOCK` constant,
+  * as noted in https://github.com/compound-finance/compound-protocol/issues/230
```
```suggestion
  * @notice Version 4 modifies the number of second per block to 12, and takes the 
  * stability fee of ETH-B as a reference.
```

fmorisan replied:
👍

MathisGD commented on contracts/DAIInterestRateModelV4.sol:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * @notice The assumed (1 - reserve factor) used to calculate the minimum borrow rate (reserve factor = 0.05)
+     */
+    uint public constant assumedOneMinusReserveFactorMantissa = 0.95e18;
```
The current reserve factor on the DAI market is 15%, so it should be this:
```suggestion
    uint public constant assumedOneMinusReserveFactorMantissa = 0.85e18;
```

fmorisan replied:
makes sense! 👍

MathisGD commented on contracts/DAIInterestRateModelV4.sol:
```diff
@@ -0,0 +1,132 @@
+     * @param owner_ The address of the owner, i.e. the Timelock contract (which has the ability to update parameters directly)
+     */
+    constructor(uint jumpMultiplierPerYear, uint kink_, address pot_, address jug_, address owner_) JumpRateModelV2(0, 0, jumpMultiplierPerYear, kink_, owner_) public {
+        gapPerBlock = 4e16 / blocksPerYear;
```
I think that the gap should be discussed. 

ETH-B stability fee + 4% = 7%, vs 5.13% at the moment.

fmorisan replied:
I don't think I get this. Should we discuss this in the governance forum?

MathisGD replied:
I meant that there is no real reason why we would keep the same as before. @monnet-supply proposed some parameters here, for example: https://www.comp.xyz/t/compound-dsr-proposal/3856/7?u=mathisgd

MathisGD reviewed (commented):
Cool @fmorisan !

One question: why didn't you modified and renamed `DAIInterestRateModelV3` ? I'm not sure that it is useful to keep the third version in the repo, and it would have make the changes clearer.

fmorisan committed:
chore: remove DAIInterestRateModelV3

fmorisan committed:
docs: change wording in DAI IRM v4 notice

fmorisan committed:
fix: change DAI reserve factor to reflect current market conditions

fmorisan commented:
Updated the contract and did some cleanup. Left a comment on one of the suggested changes. Thanks @MathisGD

MathisGD reviewed (approved):
LGTM

The community still need to choose if the DAI's IRM should be linked to the DSR on chain or not, though (see the discussion about it [here](https://www.comp.xyz/t/compound-dsr-proposal/3856/3?u=mathisgd)).

If this version is chosen, the gap may also be updated ([ref](https://github.com/compound-finance/compound-protocol/pull/231#discussion_r1055885229)).

fmorisan committed:
fix: mul before div to avoid precision loss

fmorisan committed:
docs: comment on reserve factor mantissa change

red-swan commented on contracts/DAIInterestRateModelV4.sol:
```diff
@@ -84,16 +83,16 @@ contract DAIInterestRateModelV3 is JumpRateModelV2 {
     function dsrPerBlock() public view returns (uint) {
         return (pot.dsr() - RAY_BASE) // scaled RAY_BASE aka RAY, and includes an extra "ONE" before subtraction
-            / RAY_TO_BASE_SCALE // descale to BASE
             * SECONDS_PER_BLOCK; // seconds per block
```
```suggestion
            * SECONDS_PER_BLOCK // seconds per block
```

fmorisan replied:
fixed in latest commit.

red-swan commented on contracts/DAIInterestRateModelV4.sol:
```diff
@@ -84,16 +83,16 @@ contract DAIInterestRateModelV3 is JumpRateModelV2 {
         return (pot.dsr() - RAY_BASE) // scaled RAY_BASE aka RAY, and includes an extra "ONE" before subtraction
-            / RAY_TO_BASE_SCALE // descale to BASE
             * SECONDS_PER_BLOCK; // seconds per block
+            / RAY_TO_BASE_SCALE // descale to BASE
```
```suggestion
            / RAY_TO_BASE_SCALE; // descale to BASE
```

fmorisan replied:
fixed in latest commit.

fmorisan committed:
fix: syntax

fmorisan committed:
feat: add BaseJumpRateModelV3 , fixing `blocksPerYear`

The older version of BaseJumpRateModel has a blocksPerYear constant
that reflects the amount of blocks produced in a 15-second block time environment.
After the Merge, block times have gone down from 15 seconds to 12 seconds,
this being reflected in DAIInterestRateModelV4.

This brings the constant in line with the new state of the Ethereum Network.

red-swan commented on contracts/JumpRateModelV3.sol:
```diff
@@ -0,0 +1,29 @@
+
+    constructor(uint baseRatePerYear, uint multiplierPerYear, uint jumpMultiplierPerYear, uint kink_, address owner_)
+
+    BaseJumpRateModelV2(baseRatePerYear,multiplierPerYear,jumpMultiplierPerYear,kink_,owner_) public {}
```
```suggestion
    BaseJumpRateModelV3(baseRatePerYear,multiplierPerYear,jumpMultiplierPerYear,kink_,owner_) public {}
```
